spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,333 @@
+    layer: Module[T],
+    decoder_input: ModuleNode[T],
+    encoder_outputs: ModuleNode[T],
+    attention_bias: ModuleNode[T], preName: String): ModuleNode[T] = {
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+      .inputs(decoder_input)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+      .inputs(layer.setName(preName + "/encdec_attention")
+        .inputs(norm, encoder_outputs, attention_bias))
+    CAddTable().inputs(decoder_input, drop)
+  }
+
+  private def prePostProcessingFFN(layer: Module[T],
+    decoder_input: ModuleNode[T], preName: String): ModuleNode[T] = {
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+      .inputs(decoder_input)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+      .inputs(layer.setName(preName + "/ffn").inputs(norm))
+    CAddTable().inputs(decoder_input, drop)
+  }
+}
+
+
+object TransformerLayer {
+  def apply[T: ClassTag](
+     vocabSize: Int,
+     hiddenSize: Int,
+     numHeads: Int,
+     filterSize: Int,
+     numHiddenlayers: Int,
+     postprocessDropout: Float,
+     attentionDropout: Float,
+     reluDropout: Float,
+      problem: ProblemType = LanguageModel)
+   (implicit ev: TensorNumeric[T]): TransformerLayer[T] =
+    new TransformerLayer(vocabSize, hiddenSize, numHeads,
+      filterSize, numHiddenlayers,
+      postprocessDropout, attentionDropout, reluDropout, problem)
+}
+
+private[nn] class EncodePositionConstant[T: ClassTag](implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
+  @transient private var rangeBuffer : Tensor[T] = null
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    if (!output.isEmpty && output.nElement() == input.nElement()) return output
+    val length = input.size(2)
+    val channel = input.size(3)
+
+    if (rangeBuffer == null) {
+      rangeBuffer = Tensor[T]()
+      TransformerOperation.initRangeTensor(length, rangeBuffer)
+    }
+
+    output.resize(length, channel)
+    TransformerOperation.addTimingSignal1D(length, channel,
+      rangeBuffer = rangeBuffer, timeBuffer = output)
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    if (!gradInput.isEmpty && gradInput.nElement() == input.nElement()) return gradInput
+    gradInput.resizeAs(input).zero()
+    gradInput
+  }
+}
+
+private[nn] class AttentionBiasConstant[T: ClassTag](implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    output.resizeAs(input).copy(input)
+    output = TransformerOperation.getPaddingBias(output)
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    gradInput.resizeAs(input).zero()
+    gradInput
+  }
+}
+
+private[nn] class TransformerPrepareDecoder[T: ClassTag](implicit ev: TensorNumeric[T])
```

Comment:
is this part of preprcessing ?

Reply:
yes. It is encode position.

Reply:
why not just name it `Decoder` as it's inside the same class ?

Reply:
Now, I named it `PositionEncodeWithShift` as it is shift input and then do position encode.